embassy-usb-dfu/src/application.rs

```diff
@@ -0,0 +1,135 @@
+use embassy_boot::BlockingFirmwareUpdater;
+use embassy_time::{Duration, Instant};
+use embassy_usb::control::{InResponse, OutResponse, Recipient, RequestType};
+use embassy_usb::driver::Driver;
+use embassy_usb::{Builder, Handler};
+use embedded_storage::nor_flash::NorFlash;
+
+use crate::consts::{
+    DfuAttributes, Request, State, Status, APPN_SPEC_SUBCLASS_DFU, DESC_DFU_FUNCTIONAL, DFU_PROTOCOL_RT,
+    USB_CLASS_APPN_SPEC,
+};
+
+/// Internal state for the DFU class
+pub struct Control<'d, DFU: NorFlash, STATE: NorFlash> {
+    updater: BlockingFirmwareUpdater<'d, DFU, STATE>,
+    attrs: DfuAttributes,
+    state: State,
+    timeout: Option<Duration>,
+    detach_start: Option<Instant>,
+}
+
+impl<'d, DFU: NorFlash, STATE: NorFlash> Control<'d, DFU, STATE> {
+    pub fn new(updater: BlockingFirmwareUpdater<'d, DFU, STATE>, attrs: DfuAttributes) -> Self {
+        Control {
+            updater,
+            attrs,
+            state: State::AppIdle,
+            detach_start: None,
+            timeout: None,
+        }
+    }
+}
+
+impl<'d, DFU: NorFlash, STATE: NorFlash> Handler for Control<'d, DFU, STATE> {
+    fn reset(&mut self) {
+        if let Some(start) = self.detach_start {
+            let delta = Instant::now() - start;
+            let timeout = self.timeout.unwrap();
+            #[cfg(feature = "defmt")]
+            defmt::info!(
+                "Received RESET with delta = {}, timeout = {}",
+                delta.as_millis(),
+                timeout.as_millis()
+            );
+            if delta < timeout {
+                self.updater.mark_dfu().expect("Failed to mark DFU mode in bootloader");
+                cortex_m::asm::dsb();
+                cortex_m::peripheral::SCB::sys_reset();
+            }
+        }
+    }
```

Comment:
Embassy's HALs are all cortex-m chips so I felt comfortable using that there.
One idea that comes to mind if this is desirable is to make a feature gated `sys_reset()` item that delegates to correct implementations for chips. We could also add a `custom-reset` feature with this that allows a user to pass a `() -> !` function pointer which'd get called internally.

Reply:
I was thinking of the esp32c3 which is riscv and also supported by the executor could be a potential candidate. The HAL is in the esp-rs organization, and I'm not sure if it supports the USB interface of embassy, but maybe it could.

Reply:
even on cortex-m, you might want to do custom things before reboot, like saving data or shutting down motors. 

Maybe we could have a `trait Reset`, make the code generic over it, and provide an impl for cortex-m.